app/bundles/EmailBundle/Tests/Helper/MailHelperTest.php

```diff
@@ -848,30 +852,67 @@ public function testThatTokenizationIsDisabledIfTransportDoesnotSupportTokenizat
 
     public function testImagesEmbeddedOnSend(): void
     {
-        $parameterMap = [
+        $unsubscribeUrl   = '/unsubscribe';
+        $trackingPixelUrl = '/tracking.gif';
+        $parameterMap     = [
             ['mailer_convert_embed_images', false, true],
+            ['mailer_append_tracking_pixel', false, true],
         ];
         $mockFactory = $this->getMockFactory(true, $parameterMap);
 
+        $router          = $this->createMock(UrlGeneratorInterface::class);
+        $eventDispatcher = $this->createMock(EventDispatcherInterface::class);
+        $mockFactory->expects(self::once())
+            ->method('getDispatcher')
+            ->willReturn($eventDispatcher);
+        $mockFactory->expects(self::exactly(3))
+            ->method('getRouter')
+            ->willReturn($router);
+
         $transport   = new SmtpTransport();
         $swiftMailer = new Swift_Mailer($transport);
 
         $mailer = new MailHelper($mockFactory, $swiftMailer, ['nobody@nowhere.com' => 'No Body']);
         $mailer->addTo($this->contacts[0]['email']);
-
-        $initialHtml = 'Text <img src="/tmp/fake.jpg">';
+        $mailer->setIdHash();
+
+        $initialHtml = 'Text <a href="https://mautic.com">Mautic</a> <img src="{ token }" /> <img src="/tmp/fake.jpg">';
```

Comment:
Can you add also an image with URL that is partly dynamic? Like
```suggestion
        $initialHtml = 'Text <a href="https://mautic.com">Mautic</a> <img src="{ token }" /> <img src="/tmp/fake.jpg"> <img src="https://some.url/contact/{ token2 }/avatar.png" />';
```

Reply:
I would say this will test token replacement, but suggested part is added.

Reply:
I see. It does not need to have a valid URL to encode it. I thought it is important to build the URL as you did for `{ token }` in the test. Thanks!